plutus-core/plutus-core/src/PlutusCore/Value.hs

```diff
@@ -175,23 +241,41 @@ empty :: Value
 empty = Value mempty mempty 0 0
 {-# INLINE empty #-}
 
-toList :: Value -> [(K, [(K, Integer)])]
+toList :: Value -> [(K, [(K, Quantity)])]
 toList = Map.toList . Map.map Map.toList . unpack
 {-# INLINEABLE toList #-}
 
-toFlatList :: Value -> [(K, K, Integer)]
+toFlatList :: Value -> [(K, K, Quantity)]
 toFlatList (toList -> xs) = [(c, t, a) | (c, ys) <- xs, (t, a) <- ys]
 {-# INLINEABLE toFlatList #-}
 
-fromList :: [(K, [(K, Integer)])] -> Value
-fromList =
-  pack
-    . Map.fromListWith (Map.unionWith (+))
-    . fmap (second (Map.fromListWith (+)))
+fromList :: [(K, [(K, Quantity)])] -> BuiltinResult Value
+fromList xs = do
+  -- Use unchecked addition during construction
+  let outerMap =
+        Map.fromListWith
+          (Map.unionWith unsafeAddQuantity) -- combine inner maps with unchecked addition
+          (second (Map.fromListWith unsafeAddQuantity) <$> xs)
+  -- Validate all quantities are within bounds
+  pack <$> validateNestedMap "fromList" outerMap
 {-# INLINEABLE fromList #-}
 
+-- | Unsafe addition of quantities without bounds checking.
+unsafeAddQuantity :: Quantity -> Quantity -> Quantity
+unsafeAddQuantity (UnsafeQuantity x) (UnsafeQuantity y) = UnsafeQuantity (x + y)
+{-# INLINE unsafeAddQuantity #-}
+
+-- | Validate all quantities in a nested map are within bounds.
+validateNestedMap :: String -> NestedMap -> BuiltinResult NestedMap
+validateNestedMap context =
+  traverse $ traverse \(UnsafeQuantity i) ->
```

Comment:
You shouldn't use `traverse` as it pointlessly reallocates the entire map. Just check the condition using a fold and return the original map if the check is successful or `fail` otherwise.

Reply:
Yeah, just use `any`